### PR TITLE
fix(incsearch): support `c_CTRL-{G,T}` with an offset

### LINKED ISSUE #19991: incsearch does not support `CTRL-{G,T}` with search offsets
**Is your feature request about something that is currently impossible or hard to do? Please describe the problem.**

See neovim/neovim#22228.

This is [currently documented behavior](https://github.com/vim/vim/blob/e6a84bb6b0e45ddb106f689054b0d96aba161f67/runtime/doc/cmdline.txt#L466-L471) but I see no reason why it can't be implemented.

In fact, I've made a [PR to neovim](https://github.com/neovim/neovim/pull/39097) that adds partial support for this.

**Describe the solution you'd like**

incsearch and incsearch live preview work with offsets.

**Describe alternatives you've considered**

doing nothing

**Additional context**
This issue is not fully resolved by the other PR. I still had comments that I wanted to flesh out about the highlighting [here](https://github.com/neovim/neovim/issues/22228#issuecomment-4256577488).

(Apologies if referring to neovim issues/prs here is not appropriate but I did not want to strip this request of its original circumstances).

### PR DESCRIPTION
**Pull request by barrettruth**
Closes #19991.
Ports neovim/neovim#39097.

# Problem

`c_CTRL-{G,T}` stop cycling once an `'incsearch'` command includes an offset such as `/blah/e`. `do_search()` already parses the pattern and offset separately but `may_adjust_incsearch_highlighting()` cycles on the raw cmdline text instead.

# Solution

Extract the pattern/offset parsing block from `do_search()` into a new helper `parse_search_pattern_offset()` and use it in the command-line incsearch path so `c_CTRL-{G,T}` keep selecting matches.

***NOTE***: this only fixes the cycling. The `'incsearch'` highlight still does not honor the `|search-offset|`. I was imagining this as the second half of the original report. Please see my [original comment](https://github.com/neovim/neovim/issues/22228#issuecomment-4256577488) to see the remaining behavior in question!.

### CODE REVIEW

File: src/search.c [hunk 1]
```diff
@@ -1215,6 +1215,109 @@ first_submatch(regmmatch_T *rp)
 }
 #endif
 
+/*
```
Comment (barrettruth) [inline src/search.c:1218]: entire block copy-pasted, comments included (no changes)

File: src/ex_getln.c [hunk 2]
```diff
@@ -639,31 +643,16 @@ may_adjust_incsearch_highlighting(
...
-	patlen--;
-	if (pat[patlen - 1] == '\\')
-	{
-	    pat[patlen - 1] = firstc;
-	    bslsh = TRUE;
-	}
-    }
+    (void)parse_search_pattern_offset(&pat, &patlen_s, search_delim,
+				    SEARCH_OPT, &strcopy, &searchstr,
+				    &searchstrlen, &dircp, &offset);
```
[the remark this answers is not in the document]
Comment (barrettruth) [reply src/ex_getln.c:655]: update call site to pass the remaining cmdline length into `parse_search_pattern_offset()`

File: runtime/doc/cmdline.txt [hunk 3]
```diff
@@ -466,14 +466,13 @@ CTRL-L		A match is done on the pattern in front of the cursor.  If
 	                                            *c_CTRL-G* */_CTRL-G*
 CTRL-G		When 'incsearch' is set, entering a search pattern for "/" or
 		"?" and the current match is displayed then CTRL-G will move
-		to the next match (does not take |search-offset| into account)
+		to the next match.  The |search-offset| has no effect.
```
Comment (h-east) [inline runtime/doc/cmdline.txt]: This sentence might be misleading, so how about something like this?

```suggestion
		to the next match.  The |search-offset| is applied when <CR>
		is pressed, but does not affect the preview highlighting.
```

File: runtime/doc/cmdline.txt [hunk 4]
```diff
@@ -466,14 +466,13 @@ CTRL-L		A match is done on the pattern in front of the cursor.  If
...
 		Use CTRL-T to move to the previous match.  Hint: on a regular
 		keyboard G is below T.
 	                                            *c_CTRL-T* */_CTRL-T*
 CTRL-T		When 'incsearch' is set, entering a search pattern for "/" or
 		"?" and the current match is displayed then CTRL-T will move
-		to the previous match (does not take |search-offset| into
-		account).
+		to the previous match.  The |search-offset| has no effect.
```
Comment (h-east) [inline runtime/doc/cmdline.txt]: This sentence might be misleading, so how about something like this?

```suggestion
		to the previous match.  The |search-offset| is applied when
		<CR> is pressed, but does not affect the preview highlighting.
```

**Comment by h-east** [comment]
LGTM overall.  The doc tweak reads well now.

One optional follow-up: the added tests cover forward `/` with `/e`, but
the function also handles backward `?`, negative offsets, line offsets,
and `\/` escaping inside the pattern.  A couple of additional assertions
would lock in those paths and guard against regressions:

```vim
" backward ? with /e
call feedkeys("G$?blah?e\<C-G>\<CR>", 'tx')
" /e with negative offset
call feedkeys("gg0/blah/e-2\<C-G>\<CR>", 'tx')
" pattern containing escaped slash
call feedkeys("gg0/blah\\/\<C-G>\<CR>", 'tx')
```

Not a blocker — happy to see it merged as-is if the author prefers a follow-up patch.

**Comment by barrettruth** [comment]
added some combination of those tests that i thought covered some more edge cases

ready to merge

**Comment by chrisbra** [comment]
Thanks, I also added a test for using the `+1` line offset.

**Comment by barrettruth** [comment]
brilliant!
